storage/src/main/java/tech/pegasys/teku/storage/client/EarliestAvailableBlockSlot.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright ConsenSys Software Inc., 2023
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.storage.client;
+
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.time.TimeProvider;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.storage.api.StorageQueryChannel;
+
+class EarliestAvailableBlockSlot {
+  private static final Logger LOG = LogManager.getLogger();
+  private final StorageQueryChannel historicalChainData;
+  private final TimeProvider timeProvider;
+
+  private final int earliestAvailableBlockSlotFrequency;
+
+  private UInt64 queryTime = UInt64.ZERO;
+
+  private SafeFuture<Optional<UInt64>> earliestAvailableBlockSlotFuture =
+      SafeFuture.completedFuture(Optional.empty());
+
+  public EarliestAvailableBlockSlot(
+      StorageQueryChannel historicalChainData,
+      TimeProvider timeProvider,
+      int earliestAvailableBlockSlotFrequency) {
+    this.historicalChainData = historicalChainData;
+    this.timeProvider = timeProvider;
+    this.earliestAvailableBlockSlotFrequency = earliestAvailableBlockSlotFrequency;
+  }
+
+  public SafeFuture<Optional<UInt64>> get() {
+    if (!earliestAvailableBlockSlotFuture.isDone()) {
+      // if the current query is not done, we should just return the same future
+      return earliestAvailableBlockSlotFuture;
+    }
+
+    if (queryTime
```

Comment:
I couldn't sleep with it. Why didn't you replace next 6 lines to just `actuallyGetEarliestBlockSlot();`?
You don't want fall into `synchronized` when it's not needed? But assuming most users will run it with default settings, they will fall in it and run this part of code twice

Reply:
It's basically re-using the same cached query that's already running...

Reply:
we dont even need to go into the synchronized section, because we have a future thats incomplete, so can just skip all the processing completely.

Reply:
> It's basically re-using the same cached query that's already running...
I understand it, the question was, why to calculate this twice, with default 0 value, we will count `queryTime
        .plus(earliestAvailableBlockSlotFrequency)
        .isLessThanOrEqualTo(timeProvider.getTimeInSeconds())` here and next in synchronized section the same again
```
    if (queryTime
        .plus(earliestAvailableBlockSlotFrequency)
        .isLessThanOrEqualTo(timeProvider.getTimeInSeconds())) {
      return actuallyGetEarliestBlockSlot();
    }
    return earliestAvailableBlockSlotFuture;
  }

  private synchronized SafeFuture<Optional<UInt64>> actuallyGetEarliestBlockSlot() {
    if (queryTime
        .plus(earliestAvailableBlockSlotFrequency)
        .isLessThanOrEqualTo(timeProvider.getTimeInSeconds())) {
      queryTime = timeProvider.getTimeInSeconds();
      LOG.trace("Setting earliestBlockSlot query time to {}", queryTime);
    } else {
      return earliestAvailableBlockSlotFuture;
    }
```

Reply:
There's no guarantee it hasn't happened in the meantime... and it's a trivial computation...